applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp

```diff
@@ -0,0 +1,603 @@
+        for (IndexType d = 0; d < dim; ++d) {
+#pragma omp atomic
+            r_mom[d] += rhs[i_node * block_size + (d + 1)];
+        }
+#pragma omp atomic
+        r_geometry[i_node].FastGetSolutionStepValue(REACTION_ENERGY) += rhs[i_node * block_size + 4];
+    }
+}
+
+template <>
+void CompressibleNavierStokesExplicit<2>::CalculateMassMatrix(
+    MatrixType &rMassMatrix,
+    const ProcessInfo &rCurrentProcessInfo)
+{
+    constexpr IndexType n_nodes = 3;
+    constexpr IndexType block_size = 4;
+
+    // Hardcoded shape functions for linear triangular element
+    // This is explicitly done to minimize the allocation and matrix acceses
+    // The notation N_i_j means shape function for node j in Gauss pt. i
+    const double one_sixt = 1.0 / 6.0;
+    const double two_third = 2.0 / 3.0;
+    const double N_0_0 = one_sixt;
+    const double N_0_1 = one_sixt;
+    const double N_0_2 = two_third;
+    const double N_1_0 = one_sixt;
+    const double N_1_1 = two_third;
+    const double N_1_2 = one_sixt;
+    const double N_2_0 = two_third;
+    const double N_2_1 = one_sixt;
+    const double N_2_2 = one_sixt;
+
+    // Initialize and fill the mass matrix values
+    const unsigned int size = n_nodes * block_size;
+    rMassMatrix = ZeroMatrix(size, size);
+    rMassMatrix(0, 0) = N_0_0; rMassMatrix(0, 4) = N_0_1; rMassMatrix(0, 8) = N_0_2;
+    rMassMatrix(1, 1) = N_0_0; rMassMatrix(1, 5) = N_0_1; rMassMatrix(1, 9) = N_0_2;
+    rMassMatrix(2, 2) = N_0_0; rMassMatrix(2, 6) = N_0_1; rMassMatrix(2, 10) = N_0_2;
+    rMassMatrix(3, 3) = N_0_0; rMassMatrix(3, 7) = N_0_1; rMassMatrix(3, 11) = N_0_2;
+    rMassMatrix(4, 0) = N_1_0; rMassMatrix(4, 4) = N_1_1; rMassMatrix(4, 8) = N_1_2;
+    rMassMatrix(5, 1) = N_1_0; rMassMatrix(5, 5) = N_1_1; rMassMatrix(5, 9) = N_1_2;
+    rMassMatrix(6, 2) = N_1_0; rMassMatrix(6, 6) = N_1_1; rMassMatrix(6, 10) = N_1_2;
+    rMassMatrix(7, 3) = N_1_0; rMassMatrix(7, 7) = N_1_1; rMassMatrix(7, 11) = N_1_2;
+    rMassMatrix(8, 0) = N_2_0; rMassMatrix(8, 4) = N_2_1; rMassMatrix(8, 8) = N_2_2;
+    rMassMatrix(9, 1) = N_2_0; rMassMatrix(9, 5) = N_2_1; rMassMatrix(9, 9) = N_2_2;
+    rMassMatrix(10, 2) = N_2_0; rMassMatrix(10, 6) = N_2_1; rMassMatrix(10, 10) = N_2_2;
+    rMassMatrix(11, 3) = N_2_0; rMassMatrix(11, 7) = N_2_1; rMassMatrix(11, 11) = N_2_2;
+
+    // Here we assume that all the Gauss pt. have the same weight so we multiply by the volume
+    rMassMatrix *= GetGeometry().Area();
+}
+
+template <>
+void CompressibleNavierStokesExplicit<3>::CalculateMassMatrix(
+    MatrixType &rMassMatrix,
+    const ProcessInfo &rCurrentProcessInfo)
+{
+    constexpr IndexType n_nodes = 4;
+    constexpr IndexType block_size = 5;
+
+    // Hardcoded shape functions for linear tetrahedra element
+    // This is explicitly done to minimize the alocation and matrix acceses
+    // The notation N_i_j means shape function for node j in Gauss pt. i
+    const double N_0_0 = 0.58541020;

```

Comment:
take care about this. it should be Ni*Nj (integrated)

Reply:
Copy paste error.